noir-projects/noir-protocol-circuits/crates/rollup-lib/src/abis/block_root_or_block_merge_public_inputs.nr

```diff
@@ -0,0 +1,135 @@
+use dep::types::{
+    abis::{append_only_tree_snapshot::AppendOnlyTreeSnapshot, global_variables::GlobalVariables},
+    constants::BLOCK_ROOT_OR_BLOCK_MERGE_PUBLIC_INPUTS_LENGTH,
+    partial_state_reference::PartialStateReference, traits::{Empty, Serialize, Deserialize},
+    utils::reader::Reader, address::EthAddress
+};
+use crate::abis::constant_rollup_data::ConstantRollupData;
+
+struct FeeRecipient {
+  recipient: EthAddress,
+  value: Field,
+}
+
+impl Empty for FeeRecipient {
+  fn empty() -> Self {
+    FeeRecipient {
+      recipient: EthAddress::zero(),
+      value: 0,
+    }
+  }
+}
+
+impl Serialize<2> for FeeRecipient {
+  fn serialize(self) -> [Field; 2] {
+    [self.recipient.to_field(), self.value]
+  }
+}
+
+impl Deserialize<2> for FeeRecipient {
+  fn deserialize(values: [Field; 2]) -> Self {
+    Self {
+      recipient: EthAddress::from_field(values[0]),
+      value: values[1]
+    }
+  }
+}
+
+impl Eq for FeeRecipient {
+  fn eq(self, other: Self) -> bool {
+    (self.recipient.eq(other.recipient)) & (self.value == other.value)
+  }
+}
+
+// TODO: instead of archives + global vars, use ConstantRollupData x2? It also includes vk root
+// may be confusing as new_constant.last_archive would actually be the new_archive
+struct BlockRootOrBlockMergePublicInputs {
+  previous_archive: AppendOnlyTreeSnapshot, // Archive tree root immediately before this block range
+  new_archive: AppendOnlyTreeSnapshot, // Archive tree root after adding this block range
+  previous_block_hash: Field, // Identifier of the previous block before the range
+  end_block_hash: Field, // Identifier of the last block in the range
+  start_global_variables: GlobalVariables, // Global variables for the first block in the range
+  end_global_variables: GlobalVariables, // Global variables for the last block in the range
+  out_hash: Field, // Merkle node of the L2-to-L1 messages merkle roots in the block range
+  fees: [FeeRecipient; 32], // Concatenation of all coinbase and fees for the block range
+  vk_tree_root: Field, // Root of allowed vk tree
```

Comment:
Not mentioned in engineering-designs doc - added the vk tree root to check for allowed circuits in future.